lib/Basics/SizeLimitedString.h

```diff
@@ -0,0 +1,181 @@
+
+  size_t size() const noexcept {
+    TRI_ASSERT(N >= _offset);
+    return _offset;
+  }
+
+  void push_back(char c) noexcept {
+    if (_full) {
+      return;
+    }
+    _buffer[_offset] = c;
+    ++_offset;
+    recalculateState();
+  }
+
+  SizeLimitedString& append(std::string_view data) noexcept {
+    return append(data.data(), data.size());
+  }
+
+  SizeLimitedString& append(char const* data) noexcept {
+    return append(data, strlen(data));
+  }
+
+  SizeLimitedString& append(char const* data, size_t length) noexcept {
+    if (!_full) {
+      // append as much as is possible (potentially only a prefix)
+      length = std::min(length, remaining());
+      memcpy(_buffer + _offset, data, length);
+      _offset += length;
+      recalculateState();
+    }
+    return *this;
+  }
+
+  // append a uint64_t value as a string. will only append the value if
+  // there is enough capacity in the buffer for the maximum possible
+  // uint64_t value (which is 21 bytes long).
+  SizeLimitedString& appendUInt64(uint64_t value) noexcept {
+    if (!_full) {
+      if (basics::maxUInt64StringSize > remaining()) {
+        _full = true;
+      } else {
+        _offset += basics::StringUtils::itoa(value, _buffer + _offset);
+        recalculateState();
+      }
+    }
+    return *this;
+  }
+
+  // append a hex-encoded value.
+  // will append up to as many bytes as there is remaining space in the buffer.
+  template<typename T>
+  SizeLimitedString& appendHexValue(T value, bool stripLeadingZeros) noexcept {
+    static_assert(std::is_integral_v<T> || std::is_pointer_v<T>);
+    // copy value into local buffer
+    unsigned char buffer[sizeof(T)];
+    memcpy(buffer, &value, sizeof(T));
+
+    // stringify value. if string is already full, this may do some
+    // iterations, but not append to the buffer. we are not optimizing for
+    // this case.
+    constexpr char chars[] = "0123456789abcdef";
+    unsigned char const* e = &buffer[0] + sizeof(T);
+
+    if constexpr (basics::isLittleEndian()) {
+      while (--e >= &buffer[0]) {
+        unsigned char c = *e;
+        if (!stripLeadingZeros || (c >> 4U) != 0) {
+          push_back(chars[c >> 4U]);
+          stripLeadingZeros = false;
+        }
+        if (!stripLeadingZeros || (c & 0xfU) != 0) {
+          push_back(chars[c & 0xfU]);
+          stripLeadingZeros = false;
+        }
+      }
+      if (stripLeadingZeros) {
+        push_back('0');
+      }
+    } else {
+      unsigned char const* p = &buffer[0];
+      while (p < e) {
+        unsigned char c = *p;
+        if (!stripLeadingZeros || (c >> 4U) != 0) {
+          push_back(chars[c >> 4U]);
+          stripLeadingZeros = false;
+        }
+        if (!stripLeadingZeros || (c & 0xfU) != 0) {
+          push_back(chars[c & 0xfU]);
+          stripLeadingZeros = false;
+        }
+        ++p;
+      }
+      if (stripLeadingZeros) {
+        push_back('0');
+      }
+    }
+    return *this;
```

Comment:
This can be simplified as follows:
```cpp
    std::array<unsigned char, sizeof(T)> buffer[sizeof(T)];
    memcpy(buffer, &value, sizeof(T));
    if constexpr (basics::isLittleEndian()) {
      std::reverse(buffer.begin(), buffer.end());
    }

    // stringify value. if string is already full, this may do some
    // iterations, but not append to the buffer. we are not optimizing for
    // this case.
    constexpr char chars[] = "0123456789abcdef";
    for (auto c : buffer) {
      if (!stripLeadingZeros || (c >> 4U) != 0) {
        push_back(chars[c >> 4U]);
        stripLeadingZeros = false;
      }
      if (!stripLeadingZeros || (c & 0xfU) != 0) {
        push_back(chars[c & 0xfU]);
        stripLeadingZeros = false;
      }
    }
    if (stripLeadingZeros) {
      push_back('0');
    }

    return *this;
```